src/device-manager.ts

```diff
@@ -0,0 +1,160 @@
+
+  constructor(wavRecorder: any, callbacks: DeviceManagerCallbacks, options: DeviceManagerOptions = {}) {
+    this.wavRecorder = wavRecorder;
+    this.callbacks = callbacks;
+    this.options = {
+      autoSwitchOnDisconnect: options.autoSwitchOnDisconnect ?? true,
+      listenForDeviceChanges: options.listenForDeviceChanges ?? true,
+    };
+
+    this.updateCurrentDeviceId();
+
+    if (this.options.listenForDeviceChanges) {
+      this._setupDeviceChangeListener();
+    }
+  }
+
+  /**
+   * Updates the current device ID from the active audio stream
+   */
+  updateCurrentDeviceId(): void {
+    try {
+      const currentStream = this.wavRecorder.getStream();
+      const currentTrack = currentStream?.getAudioTracks()[0];
+      if (currentTrack) {
+        this.currentDeviceId = currentTrack.getSettings().deviceId || null;
+      }
+    } catch (error) {
+      // Silent fail - device ID will be null
+    }
+  }
+
+  /**
+   * Handles errors consistently
+   */
+  private handleError(message: string, error: unknown): Error {
+    const err = error instanceof Error ? error : new Error(String(error));
+    console.error(message, err);
+    this.callbacks.onDeviceError(err);
+    return err;
+  }
+
+  /**
+   * Sets up the device change event listener
+   */
+  private _setupDeviceChangeListener(): void {
+    this.deviceChangeListener = async (devices: MediaDeviceWithDefault[]) => {
+      try {
+        // Ensure we have current device ID
+        if (!this.currentDeviceId) {
+          this.updateCurrentDeviceId();
+        }
+
+        // Check if current device was disconnected
+        const currentDeviceExists = devices.some((device) => device.deviceId === this.currentDeviceId);
+
+        if (!currentDeviceExists && this.currentDeviceId) {
+          // Device was disconnected
+          if (this.callbacks.onDeviceDisconnected) {
+            this.callbacks.onDeviceDisconnected(this.currentDeviceId);
+          }
+
+          // Auto-switch if enabled
+          if (this.options.autoSwitchOnDisconnect) {
+            await this.switchToNextDevice();
+          }
+        }
+      } catch (error) {
+        this.handleError('Error handling device change:', error);
+      }
+    };
+
+    this.wavRecorder.listenForDeviceChange(this.deviceChangeListener);
+  }
+
+  /**
+   * Switches to the next available device
+   */
+  async switchToNextDevice(): Promise<void> {
+    try {
+      const devices = await this.wavRecorder.listDevices();
+      const audioInputs = devices.filter((d: MediaDeviceWithDefault) => d.kind === 'audioinput');
+
+      // Find first device that's not the current one
+      const nextDevice = audioInputs.find((d: MediaDeviceWithDefault) => d.deviceId !== this.currentDeviceId);
+
+      if (nextDevice) {
+        const fromDeviceId = this.currentDeviceId;
+        await this.setInputDevice(nextDevice.deviceId);
+
+        if (this.callbacks.onDeviceSwitched && fromDeviceId) {
```

Comment:
Should we move this be within setInputDevice, so that the cb is called whenever the input device changes by any means?